helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,310 @@
+        .for_each(|render| render(context, write_center));
+
+    // Width of the empty space between the left and center area and between the center and right area.
+    let spacing = 1u16;
+
+    let edge_width = context.parts.left.width().max(context.parts.right.width()) as u16;
+    let center_max_width = viewport.width.saturating_sub(2 * edge_width + 2 * spacing);
+    let center_width = center_max_width.min(context.parts.center.width() as u16);
+
+    surface.set_spans(
+        viewport.x + viewport.width / 2 - center_width / 2,
+        viewport.y,
+        &context.parts.center,
+        center_width,
+    );
+}
+
+fn append(buffer: &mut Spans, text: String, base_style: &Style, style: Option<Style>) {
+    buffer.0.push(Span::styled(
+        text,
+        style.map_or(*base_style, |s| (*base_style).patch(s)),
+    ));
+}
+
+fn get_render_function<F>(element_id: StatusLineElementID) -> impl Fn(&mut RenderContext, F)
+where
+    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+{
+    match element_id {
+        helix_view::editor::StatusLineElement::Mode => render_mode,
+        helix_view::editor::StatusLineElement::Spinner => render_lsp_spinner,
+        helix_view::editor::StatusLineElement::FileName => render_file_name,
+        helix_view::editor::StatusLineElement::FileEncoding => render_file_encoding,
+        helix_view::editor::StatusLineElement::FileType => render_file_type,
+        helix_view::editor::StatusLineElement::Diagnostics => render_diagnostics,
+        helix_view::editor::StatusLineElement::Selections => render_selections,
+        helix_view::editor::StatusLineElement::Position => render_position,
+    }
+}
+
+fn render_mode<F>(context: &mut RenderContext, write: F)
+where
+    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+{
+    let visible = context.focused;
+
+    write(
+        context,
+        format!(
+            " {} ",
+            if visible {
+                match context.doc.mode() {
+                    Mode::Insert => "INS",
+                    Mode::Select => "SEL",
+                    Mode::Normal => "NOR",
+                }
+            } else {
+                // If not focused, explicitly leave an empty space instead of returning None.
+                "   "
+            }
+        ),
+        if visible && context.editor.config().color_modes {
+            match context.doc.mode() {
+                Mode::Insert => Some(context.editor.theme.get("ui.statusline.insert")),
+                Mode::Select => Some(context.editor.theme.get("ui.statusline.select")),
+                Mode::Normal => Some(context.editor.theme.get("ui.statusline.normal")),
+            }
+        } else {
+            None
+        },
+    );
+}
+
+fn render_lsp_spinner<F>(context: &mut RenderContext, write: F)
+where
+    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+{
+    write(
+        context,
+        format!(
+            " {} ",
```

Comment:
There seems to be additional spacing between `NOR` and the spinner now (and also on the right side).

I think the spinner should have no padding. We can also add a spacer element that simply renders an empty space so that users can space out the components as they wish

Reply:
You'll also be able to add spacing within a custom spinner with #2793.

Reply:
Oh you're right, seems like I did change the padding compared to before this PR.
I changed it back in dba88ba.

With spinner padding:
![image](https://user-images.githubusercontent.com/2804556/179393794-61fe9072-29d1-4096-96a5-24d9c8f824c8.png)

Without spinner padding (master):
![image](https://user-images.githubusercontent.com/2804556/179394110-5ea1ffd6-598a-4888-bc8d-c5186576ca50.png)

I must've introduced the padding intuitively, as it made more sense to me.